File: cpp/menu/menu.cpp
#include "menu.h"

MenuItem::MenuItem(const char* text, fptr callback):
  text(text), 
  callback(callback)
{}

const char* MenuItem::getText()
{
  return text;
}

fptr MenuItem::getCallback()
{
  return callback;
}  


Menu::Menu() 
{
  index = 0;
  n_items = 0;
  MenuStream = NULL;
}

void Menu::begin(Stream &s)
{
  MenuStream = &s;
}

void Menu::addItem(const char* text, fptr callback)
{
  if(n_items < 10) {
    items[n_items] = new MenuItem(text, callback);
    n_items++;
  }
}

void Menu::up()
{
  if(index < n_items-1){
    index++;
    if ( MenuStream != NULL )
      MenuStream->println(items[index]->getText());
  }
}

void Menu::down()
{
  if(index > 0){  
    index--;
    if ( MenuStream != NULL )
      MenuStream->println(items[index]->getText());
  }
}

void Menu::enter()
{
  fptr callback = items[index]->getCallback();

  if(callback != NULL) callback();
}

File: python/menu/menu.py
class Menu:
    def __init__(self):
        self.items = []
        self.index = 0
        self.max_items = 10
        self.num_items = 0

    def enter(self):
        action = self.items[self.index]['action']
        if action is not None:  
            action() 

    def up(self):
        if self.index < (self.num_items - 1):
            self.index += 1
            print(self.items[self.index]['text'])

    def down(self):
        if self.index > 0:
            self.index -= 1
            print(self.items[self.index]['text'])

    def addItem(self, text, action):
        if self.num_items < (self.max_items - 1):
            self.items.append({'text':text, 
                        'action':action})
            self.num_items += 1

    def printMenu(self):
        for item in self.items:
            print(item['text'])